src/client/datascience/jupyter/kernels/helpers.ts

```diff
@@ -1774,3 +1807,201 @@ export async function executeSilently(session: IJupyterSession, code: string): P
 
     return outputs;
 }
+async function switchController(
+    resource: Resource,
+    serviceContainer: IServiceContainer
+): Promise<VSCodeNotebookController | undefined> {
+    const commandManager = serviceContainer.get<ICommandManager>(ICommandManager);
+    const notebooks = serviceContainer.get<IVSCodeNotebook>(IVSCodeNotebook);
+    const editor = findNotebookEditor(resource, notebooks, serviceContainer.get(IInteractiveWindowProvider));
+
+    // Listen for selection change events (may not fire if user cancels)
+    const controllerManager = serviceContainer.get<INotebookControllerManager>(INotebookControllerManager);
+    let controller: VSCodeNotebookController | undefined;
+    const waitForSelection = createDeferred<VSCodeNotebookController>();
+    const disposable = controllerManager.onNotebookControllerSelected((e) => waitForSelection.resolve(e.controller));
+
+    const selected = (await selectKernel(
+        resource,
+        notebooks,
+        serviceContainer.get(IInteractiveWindowProvider),
+        commandManager
+    )) as boolean;
```

Comment:
I'd prefer to avoid the unnecessary casting, i don' tthink its necessary

Reply:
Yeah not sure why this is here. Seems unnecessary. I'll remove it.